### PR TITLE
DTrace, cmon, cleanup, retry downstairs connections at 10 seconds.

### PR DESCRIPTION
**Pull request by leftwo**
This is kind of a grab-bag of things I stumbled into while trying to debug some `LiveRepair` double failures
and how we might better handle them.

The only load bearing thing here is really a change in how long we wait before retrying a connection to
a disconnected downstairs.   I've bumped it up to 10 seconds, but we can further discuss values if the jump
from 1 to 10 concerns people.

The internal counters for `reconcile_repair_needed` and `reconcile_repaired` got lost at some point in 
the past.  This adds them back. 
Added a count of reconciliation aborted to the DTrace Arg struct. 
Added the extent_limit to the DTrace Arg struct.

Replaced an assert with an if/panic to print more information if we do encounter the situation.

Gather and log a more accurate count of jobs that we moved to replay.

Print fewer messages during reconciliation.

Added a 10 second delay before we reconnect to a downstairs.

Overhauled the cmon program a bit.  Their is a `dtrace` option to cmon that takes the
output from `tools/dtrace/upstairs_raw.d` and pretty prints it.  You can select which
fields from the dtrace output you want on the fly, which is a little easier than modifying
the dtrace script each time you want to change your output.
Here, we have added more of the values in the DTrace Arg and print them out.  Added a 
short version of the downstairs state to allow for narrow terminal printing.  The values
added are extent_limit, backpressure, next_job_id, job_id_delta, ds_delay, and an io_summary.

Here is some sample output from `cmon`

The first showing a similiar one to `upstairs_info.d`:
```
local:pflush$ sudo dtrace -s ./tools/dtrace/upstairs_raw.d |./target/release/cmon dtrace -o state -o job-delta -o backpressure -o ds-count -o io-summary 
dtrace: system integrity protection is on, some features will not be available

 DS0 DS1 DS2 DELTA BAKPR   DSW  NEW0  NEW1  NEW2   IP0   IP1   IP2    D0    D1    D2    S0    S1    S2
 ACT ACT ACT 27132     0    16     0     0     0     1     1     1    15    15    15     0     0     0
 ACT ACT ACT   104     0    36     0     0     0     1     1     1    35    35    35     0     0     0
 ACT ACT ACT   101     0     8     0     0     0     1     1     1     7     7     7     0     0     0
 ACT ACT ACT   124     0    13     0     0     0     1     1     1    12    12    12     0     0     0
 ACT ACT ACT   121     0     8     0     0     0     1     1     1     7     7     7     0     0     0
 ACT ACT ACT   111     0     5     0     0     0     1     4     1     4     1     4     0     0     0
 ACT ACT ACT    87     0     4     0     0     0     1     0     1     3     4     3     0     0     0
 ACT ACT ACT   123     0    19     0     0     0     1     1    10    18    18     9     0     0     0
 ACT ACT ACT   113     0    19     0     0     0     1     1     1    18    18    18     0     0     0
 ACT ACT ACT   121     0    14     0     0     0    13     1     1     1    13    13     0     0     0
 ACT ACT ACT   116     0     9     0     0     0     1     1     1     8     8     8     0     0     0
```

Here is another showing live repair in action.
First going `Offline`, then `Offline` to `Faulted`, then starting the `LiveRepair`:
```
local:pflush$ sudo dtrace -s ./tools/dtrace/upstairs_raw.d |./target/release/cmon dtrace -o state -o job-delta -o ds-count -o live-repair -o extent-live-repair -o extent-limit 
 DS0 DS1 DS2 DELTA   DSW LRC0 LRC1 LRC0 LRA0 LRA1 LRA2 EXR0 EXR1 EXR2 EXC0 EXC1 EXC2 EXTL
 ACT ACT ACT    55 12904    0    1    0    0    0    0    0    0    0    0    0    0    0
 ACT ACT ACT    62 10727    0    1    0    0    0    0    0    0    0    0    0    0    0
 ACT ACT ACT    60 10609    0    1    0    0    0    0    0    0    0    0    0    0    0
 ACT OFF ACT    54 10563    0    1    0    0    0    0    0    0    0    0    0    0    0
 ACT OFF ACT   117 10680    0    1    0    0    0    0    0    0    0    0    0    0    0
...
 DS0 DS1 DS2 DELTA   DSW LRC0 LRC1 LRC0 LRA0 LRA1 LRA2 EXR0 EXR1 EXR2 EXC0 EXC1 EXC2 EXTL
 ACT OFF ACT    65 52948    0    1    0    0    0    0    0    0    0    0    0    0    0
 ACT OFF ACT    79 53027    0    1    0    0    0    0    0    0    0    0    0    0    0
 ACT OFF ACT    62 53089    0    1    0    0    0    0    0    0    0    0    0    0    0
 ACT OFF ACT    70 53159    0    1    0    0    0    0    0    0    0    0    0    0    0
 ACT FLT ACT    25    16    0    1    0    0    0    0    0    0    0    0    0    0    0
 ACT FLT ACT    96    33    0    1    0    0    0    0    0    0    0    0    0    0    0
 ACT FLT ACT    58    14    0    1    0    0    0    0    0    0    0    0    0    0    0
...
 DS0 DS1 DS2 DELTA   DSW LRC0 LRC1 LRC0 LRA0 LRA1 LRA2 EXR0 EXR1 EXR2 EXC0 EXC1 EXC2 EXTL
 ACT FLT ACT    79    10    0    1    0    0    0    0    0    0    0    0    0    0    0
 ACT FLT ACT    98    51    0    1    0    0    0    0    0    0    0    0    0    0    0
 ACT  LR ACT    66     4    0    1    0    0    0    0    0    5    0    0    0    0    4
 ACT  LR ACT    75    10    0    1    0    0    0    0    0   11    0    0    0    0   11
 ACT  LR ACT   104    48    0    1    0    0    0    0    0   18    0    0    0    0   17
 ACT  LR ACT    71     5    0    1    0    0    0    0    0   23    0    0    0    0   22
 ACT  LR ACT    77     9    0    1    0    0    0    0    0   29    0    0    0    0   28
 ACT  LR ACT    86    27    0    1    0    0    0    0    0   35    0    0    0    0   35
 ACT  LR ACT    79    16    0    1    0    0    0    0    0   42    0    0    0    0   42
 ACT  LR ACT    79    44    0    1    0    0    0    0    0   49    0    0    0    0   49
 ACT  LR ACT    84    40    0    1    0    0    0    0    0   56    0    0    0    0   56
```

Here is `LiveRepair` wrapping up:
```
local:pflush$ sudo dtrace -s ./tools/dtrace/upstairs_raw.d |./target/release/cmon dtrace -o state -o job-delta -o ds-count -o live-repair -o extent-live-repair -o extent-limit 
 DS0 DS1 DS2 DELTA   DSW LRC0 LRC1 LRC0 LRA0 LRA1 LRA2 EXR0 EXR1 EXR2 EXC0 EXC1 EXC2 EXTL
 ACT  LR ACT    74     8    0    0    0    0    0    0    0  371    0    0    0    0  370
 ACT  LR ACT    33     8    0    0    0    0    0    0    0  375    0    0    0    0  374
 ACT  LR ACT   103    45    0    0    0    0    0    0    0  380    0    0    0    0  380
 ACT  LR ACT    66    16    0    0    0    0    0    0    0  386    0    0    0    0  385
 ACT  LR ACT   102    22    0    0    0    0    0    0    0  390    0    0    0    0  390
 ACT  LR ACT    91     6    0    0    0    0    0    0    0  395    0    0    0    0  394
 ACT ACT ACT    56     5    0    1    0    0    0    0    0  400    0    0    0    0    0
 ACT ACT ACT   109     2    0    1    0    0    0    0    0  400    0    0    0    0    0
 ACT ACT ACT   104    16    0    1    0    0    0    0    0  400    0    0    0    0    0
```

### CODE REVIEW

File: upstairs/src/client.rs [hunk 1]
```diff
@@ -443,21 +443,25 @@ impl DownstairsClient {
         out
     }
 
-    pub(crate) fn replay_job(&mut self, job: &mut DownstairsIO) {
+    pub(crate) fn replay_job(&mut self, job: &mut DownstairsIO) -> bool {
```
Comment (mkeeter) [inline upstairs/src/client.rs:446]: Add a docstring here indicating the significance of the returned `bool`, e.g.
```rust
/// Ensures that the given job is in the job queue and in `IOState::New`
///
/// Returns `true` if the job was requeued, or `false` if it was already `New`

File: upstairs/src/client.rs [hunk 2]
```diff
@@ -559,8 +563,6 @@ impl DownstairsClient {
                 self.log,
                 "Gone missing, transition from {current:?} to {new_state:?}"
             );
-        } else {
```
Comment (mkeeter) [inline upstairs/src/client.rs:562]: Was removing this intentional?

File: upstairs/src/downstairs.rs [hunk 3]
```diff
@@ -2019,6 +2031,8 @@ impl Downstairs {
 
         if self.clients[client_id].on_reconciliation_job_done(repair_id, next) {
             self.reconcile_current_work = None;
+            self.reconcile_repair_needed -= 1;
```
Comment (mkeeter) [inline upstairs/src/downstairs.rs:2034]: Were we not handling `self.reconcile_repair_needed` correctly before?

File: upstairs/src/downstairs.rs [hunk 4]
```diff
@@ -5837,6 +5859,7 @@ pub(crate) mod test {
     async fn reconcile_repair_workflow_1() {
         let mut ds = Downstairs::test_default();
         set_all_reconcile(&mut ds);
+        ds.reconcile_repair_needed = 2;
```
Comment (mkeeter) [inline upstairs/src/downstairs.rs:5862]: Nit: move this to after the calls to `push_back` and make it
```rust
ds.reconcile_repair_needed = ds.reconcile_task_list.len()
```
to make the logic clear

File: upstairs/src/downstairs.rs [hunk 5]
```diff
@@ -5886,13 +5909,16 @@ pub(crate) mod test {
...
+        assert_eq!(ds.reconcile_repair_needed, 0);
+        assert_eq!(ds.reconcile_repaired, 2);
     }
 
     #[tokio::test]
     async fn reconcile_repair_workflow_2() {
         // Verify Done or Skipped works when checking for a complete repair
         let mut ds = Downstairs::test_default();
         set_all_reconcile(&mut ds);
+        ds.reconcile_repair_needed = 1;
```
Comment (mkeeter) [inline upstairs/src/downstairs.rs:5921]: Same thing here

File: upstairs/src/lib.rs [hunk 6]
```diff
@@ -1632,6 +1634,8 @@ pub struct Arg {
     pub ds_extents_repaired: [usize; 3],
     /// Times we have live confirmed  an extent on this downstairs.
     pub ds_extents_confirmed: [usize; 3],
+    /// If in Live Repair, the current extent we are repairing.
+    pub ds_extent_limit: usize,
```
Comment (mkeeter) [inline upstairs/src/lib.rs:1638]: How close are we to the `strsize=1k` DTrace limit?

File: upstairs/src/downstairs.rs (hunk 3 above)
Comment (leftwo) [reply upstairs/src/downstairs.rs:2034]: No, it got dropped in the great one task refactor and I failed to notice its absence.

File: upstairs/src/lib.rs (hunk 6 above)
Comment (leftwo) [reply upstairs/src/lib.rs:1638]: Looks like we are currently at 843, at least that is how much JSON I get.

File: upstairs/src/client.rs [hunk 7]
```diff
@@ -443,21 +443,25 @@ impl DownstairsClient {
...
         }
+
         let old_state = self.set_job_state(job, IOState::New);
         job.replay = true;
         if old_state != IOState::New {
             self.requeue_one(job.ds_id);
+            return true;
         }
+
+        false
```
[the remark this answers is not in the document]
Comment (leftwo) [reply upstairs/src/client.rs]: Updated in 906d5b9

File: upstairs/src/client.rs (hunk 1 above)
Comment (leftwo) [reply upstairs/src/client.rs:432]: Updated in 906d5b9

File: upstairs/src/downstairs.rs (hunk 4 above)
Comment (leftwo) [reply upstairs/src/downstairs.rs]: Made clear in 906d5b9

File: upstairs/src/downstairs.rs (hunk 5 above)
Comment (leftwo) [reply upstairs/src/downstairs.rs]: Made clear in 906d5b9

File: upstairs/src/downstairs.rs (hunk 3 above)
Comment (mkeeter) [reply upstairs/src/downstairs.rs:2014]: Got it, thanks for bringing it back!

File: upstairs/src/lib.rs (hunk 6 above)
Comment (mkeeter) [reply upstairs/src/lib.rs:1630]: Great, that's not too close to the limit

**Review by mkeeter — APPROVED** [review]

**Pull request merged**